With growing demands and cut-throat competition in the market, a Superstore Giant is seeking your knowledge in understanding what works best for them. They would like to understand which products, regions, categories and customer segments they should target or avoid. <br>
You can even take this a step further and try and build a Regression model to predict Sales or Profit.<br>
Go crazy with the dataset, but also make sure to provide some business insights to improve.<br>

Metadata<br>
Row ID => Unique ID for each row.<br>
Order ID => Unique Order ID for each Customer.<br>
Order Date => Order Date of the product.<br>
Ship Date => Shipping Date of the Product.<br>
Ship Mode=> Shipping Mode specified by the Customer.<br>
Customer ID => Unique ID to identify each Customer.<br>
Customer Name => Name of the Customer.<br>
Segment => The segment where the Customer belongs.<br>
Country => Country of residence of the Customer.<br>
City => City of residence of of the Customer.<br>
State => State of residence of the Customer.<br>
Postal Code => Postal Code of every Customer.<br>
Region => Region where the Customer belong.<br>
Product ID => Unique ID of the Product.<br>
Category => Category of the product ordered.<br>
Sub-Category => Sub-Category of the product ordered.<br>
Product Name => Name of the Product<br>
Sales => Sales of the Product.<br>
Quantity => Quantity of the Product.<br>
Discount => Discount provided.<br>
Profit => Profit/Loss incurred.<br>

In [1]:
# Import dependencies
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import numpy as np
import seaborn as sns
import dtale as d 
from IPython.display import display

In [40]:
# Import dataset
df = pd.read_csv(r"C:\Users\BIOLA\Desktop\Data Science Projects\Team Code\Dataset\Superstore data.csv" , encoding='cp1252')

In [206]:
# df.drop(['Row ID'], axis = 1, inplace = True)
df.shape

(9994, 20)

In [43]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   object 
 2   Order Date     9994 non-null   object 
 3   Ship Date      9994 non-null   object 
 4   Ship Mode      9994 non-null   object 
 5   Customer ID    9994 non-null   object 
 6   Customer Name  9994 non-null   object 
 7   Segment        9994 non-null   object 
 8   Country        9994 non-null   object 
 9   City           9994 non-null   object 
 10  State          9994 non-null   object 
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   object 
 13  Product ID     9994 non-null   object 
 14  Category       9994 non-null   object 
 15  Sub-Category   9994 non-null   object 
 16  Product Name   9994 non-null   object 
 17  Sales          9994 non-null   float64
 18  Quantity

Notes from the original dataset before cleaning
- There are no missing data
- 2 columns and 9994 rows 

In [60]:
# Convert Date column to pd.Datetime
df['Order Date'] = pd.to_datetime(df['Order Date'])
df['Ship Date'] = pd.to_datetime(df['Ship Date'])
ds = df.copy()

In [62]:
ds.describe()

Postal Code         Sales     Quantity     Discount       Profit
count   9994.000000   9994.000000  9994.000000  9994.000000  9994.000000
mean   55190.379428    229.858001     3.789574     0.156203    28.656896
std    32063.693350    623.245101     2.225110     0.206452   234.260108
min     1040.000000      0.444000     1.000000     0.000000 -6599.978000
25%    23223.000000     17.280000     2.000000     0.000000     1.728750
50%    56430.500000     54.490000     3.000000     0.200000     8.666500
75%    90008.000000    209.940000     5.000000     0.200000    29.364000
max    99301.000000  22638.480000    14.000000     0.800000  8399.976000

---------------------------------
## Exploratory Data Analysis

- To-Do

which region sells the most<br>
which product sells the most<br>
which category should have more focus or avoid<br>
who is the top best customers<br>
Product that are selling at a loss <br>
State that makes the most sales <br>
Predict Sales

In [116]:
ds.head(3)

Order ID Order Date  Ship Date     Ship Mode Customer ID  \
0  CA-2016-152156 2016-11-08 2016-11-11  Second Class    CG-12520   
1  CA-2016-152156 2016-11-08 2016-11-11  Second Class    CG-12520   
2  CA-2016-138688 2016-06-12 2016-06-16  Second Class    DV-13045   

     Customer Name    Segment        Country         City       State  \
0      Claire Gute   Consumer  United States    Henderson    Kentucky   
1      Claire Gute   Consumer  United States    Henderson    Kentucky   
2  Darrin Van Huff  Corporate  United States  Los Angeles  California   

   Postal Code Region       Product ID         Category Sub-Category  \
0        42420  South  FUR-BO-10001798        Furniture    Bookcases   
1        42420  South  FUR-CH-10000454        Furniture       Chairs   
2        90036   West  OFF-LA-10000240  Office Supplies       Labels   

                                        Product Name   Sales  Quantity  \
0                  Bush Somerset Collection Bookcase  261.96         2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.94         3   
2  Self-Adhesive Address Labels for Typewriters b...   14.62         2   

   Discount    Profit  
0       0.0   41.9136  
1       0.0  219.5820  
2       0.0    6.8714

In [147]:
ds.columns

Index(['Order ID', 'Order Date', 'Ship Date', 'Ship Mode', 'Customer ID',
       'Customer Name', 'Segment', 'Country', 'City', 'State', 'Postal Code',
       'Region', 'Product ID', 'Category', 'Sub-Category', 'Product Name',
       'Sales', 'Quantity', 'Discount', 'Profit'],
      dtype='object')

In [192]:
# Which Region Sells the most
reg = ds.Sales.groupby(df.Region).sum().sort_values(ascending = False)
reg.to_frame()

Sales
Region              
West     725457.8245
East     678781.2400
Central  501239.8908
South    391721.9050

<u> The West Region sells the most

In [194]:
# What Product Sells the most
prod = ds['Sales'].groupby(ds['Product Name']).sum().sort_values(ascending = False)
prod.to_frame().head(5)

Sales
Product Name                                                 
Canon imageCLASS 2200 Advanced Copier               61599.824
Fellowes PB500 Electric Punch Plastic Comb Bind...  27453.384
Cisco TelePresence System EX90 Videoconferencin...  22638.480
HON 5400 Series Task Chairs for Big and Tall        21870.576
GBC DocuBind TL300 Electric Binding System          19823.479

<u>Above are the top 5 best selling products

In [188]:
# which category should have more focus or avoid
cat = ds.Sales.groupby(ds.Category).sum().sort_values(ascending = False)
display(cat.to_frame())
display(ds.Category.value_counts())

Sales
Category                    
Technology       836154.0330
Furniture        741999.7953
Office Supplies  719047.0320

Office Supplies    6026
Furniture          2121
Technology         1847
Name: Category, dtype: int64

<u>Although we could see that <b>*Office Supply*</b> makes the most transaction, but <b>*Technology*</b> makes most sales

In [314]:
# who is the top best customers
c = ['Customer Name', 'Sales']
c = ds[c]
c.sort_values(by= 'Sales', ascending= False).head(5)

Customer Name      Sales
2697   Sean Miller  22638.480
6826  Tamara Chand  17499.950
8153  Raymond Buch  13999.960
2623  Tom Ashbrook  11199.968
4190  Hunter Lopez  10499.970

<u>The Top Best Customer Goes to <b>Sean Miller</b> who has 25043.050 worth of purchases

In [272]:
# Products that are making a loss
loss = ds.Profit.groupby(ds['Product Name']).sum().sort_values(ascending = True)
loss.to_frame().head(10)

Profit
Product Name                                                 
Cubify CubeX 3D Printer Double Head Print          -8879.9704
Lexmark MX611dhe Monochrome Laser Printer          -4589.9730
Cubify CubeX 3D Printer Triple Head Print          -3839.9904
Chromcraft Bull-Nose Wood Oval Conference Table... -2876.1156
Bush Advantage Collection Racetrack Conference ... -1934.3976
GBC DocuBind P400 Electric Binding System          -1878.1662
Cisco TelePresence System EX90 Videoconferencin... -1811.0784
Martin Yale Chadless Opener Electric Letter Opener -1299.1836
Balt Solid Wood Round Tables                       -1201.0581
BoxOffice By Design Rectangular and Half-Moon M... -1148.4375

<u>Above are the list of Product that are selling at a loss.

In [274]:
# States the makes the most Sales
st = ds.Sales.groupby(ds.State).sum().sort_values(ascending = False)
st.to_frame().head(5)

Sales
State                    
California    457687.6315
New York      310876.2710
Texas         170188.0458
Washington    138641.2700
Pennsylvania  116511.9140

<u>California makes the most sales<br>
<u>No state sells at a loss

---------------------------------------------------
### Sales Prediction and Modelling

In [313]:
# Import Dependencies
from sklearn.model_selection import cross_val_score, GridSearchCV, RandomizedSearchCV
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error
from sklearn.feature_selection import SelectKBest, f_regression
from sklearn.feature_selection import chi2
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
import math
from sklearn.model_selection import KFold

In [280]:
# Drop the non-predictive columns
X = ds.copy() # Copy the dataset
X.drop(['Country', 'City', 'Customer Name', 'Customer ID', 'Order Date', 'Ship Date', 'Postal Code', 'Product Name', 'Product ID'], axis = 1, inplace = True)
y = X.pop('Sales') # Pick out the target column
X.shape

(9994, 9)

In [281]:
# Display Categorical Columns
X.select_dtypes(include=['object']).columns.tolist()

['Ship Mode', 'Segment', 'State', 'Region', 'Category', 'Sub-Category']

In [282]:
# Make a list of the categorical columns
cat_cols = ['Ship Mode', 'Segment', 'State', 'Region', 'Category', 'Sub-Category']

In [283]:
# Label Encode Categorical columns
lb = LabelEncoder() # Instantiate the library
for i in cat_cols:
    X[i] = lb.fit_transform(X[[i]])
X.shape

(9994, 9)

In [284]:
# Display new Dataset
X

Ship Mode  Segment  State  Region  Category  Sub-Category  Quantity  \
0             2        0     15       2         0             4         2   
1             2        0     15       2         0             5         3   
2             2        1      3       3         1            10         2   
3             3        0      8       2         0            16         5   
4             3        0      8       2         1            14         2   
...         ...      ...    ...     ...       ...           ...       ...   
9989          2        0      8       2         0             9         3   
9990          3        0      3       3         0             9         2   
9991          3        0      3       3         2            13         2   
9992          3        0      3       3         1            12         4   
9993          2        0      3       3         1             1         2   

      Discount    Profit  
0         0.00   41.9136  
1         0.00  219.5820  
2         0.00    6.8714  
3         0.45 -383.0310  
4         0.20    2.5164  
...        ...       ...  
9989      0.20    4.1028  
9990      0.00   15.6332  
9991      0.20   19.3932  
9992      0.00   13.3200  
9993      0.00   72.9480  

[9994 rows x 9 columns]

#### Train Test Split

In [305]:
X_train, X_test, y_train, y_test = train_test_split(X,y, train_size = 0.8,  random_state= 2)
print(f'X_train: {X_train.shape}')
print(f'X_test: {X_test.shape}')
print('y_train: {}'.format(y_train.shape))
print('y_test: {}'.format(y_test.shape))

X_train: (7995, 9)
X_test: (1999, 9)
y_train: (7995,)
y_test: (1999,)


#### Test for Normality and Standardization 

In [306]:
X.describe()

Ship Mode      Segment        State       Region     Category  \
count  9994.000000  9994.000000  9994.000000  9994.000000  9994.000000   
mean      2.235041     0.658995    22.343206     1.570642     0.972584   
std       1.100880     0.762622    15.504611     1.162329     0.629544   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       2.000000     0.000000     4.000000     1.000000     1.000000   
50%       3.000000     0.000000    27.000000     1.000000     1.000000   
75%       3.000000     1.000000    36.000000     3.000000     1.000000   
max       3.000000     2.000000    48.000000     3.000000     2.000000   

       Sub-Category     Quantity     Discount       Profit  
count   9994.000000  9994.000000  9994.000000  9994.000000  
mean       7.590454     3.789574     0.156203    28.656896  
std        5.051429     2.225110     0.206452   234.260108  
min        0.000000     1.000000     0.000000 -6599.978000  
25%        3.000000     2.000000     0.000000     1.728750  
50%        9.000000     3.000000     0.200000     8.666500  
75%       12.000000     5.000000     0.200000    29.364000  
max       16.000000    14.000000     0.800000  8399.976000

From the mean and standard variation of the distribution above, we can agree that only the *Profit* feature needs being normalized

In [298]:
y.to_frame().describe()

Sales
count   9994.000000
mean     229.858001
std      623.245101
min        0.444000
25%       17.280000
50%       54.490000
75%      209.940000
max    22638.480000

By good practice, its better to leave the test dataset unscaled.

In [292]:
# Test for Normalization
plt.figure(figsize=(18,2))
plt.hist(X['Profit'])
plt.show()

In [320]:
# Normalize the 'Profit' feature
min_max = MinMaxScaler() # instantiate the dependency
X_train['Profit'] = min_max.fit_transform(X_train[['Profit']])
X_test['Profit'] =  min_max.fit_transform(X_test[['Profit']])

In [323]:
# display to check the Profit normalization
X_train

Ship Mode  Segment  State  Region  Category  Sub-Category  Quantity  \
1545          2        1     30       1         0             4         7   
6048          3        1      8       2         1             3         5   
8810          3        1      3       3         1            12         4   
5157          2        2     11       0         0             5         3   
524           0        0      3       3         2             0         2   
...         ...      ...    ...     ...       ...           ...       ...   
1099          0        2      3       3         0             9         3   
2514          3        1     11       0         0             5         2   
6637          2        0     30       1         1             2         3   
2575          3        1      9       2         1             3         5   
7336          3        0     30       1         0             4         2   

      Discount    Profit  
1545       0.2  0.453112  
6048       0.7  0.371230  
8810       0.0  0.441201  
5157       0.3  0.428077  
524        0.0  0.441811  
...        ...       ...  
1099       0.0  0.440273  
2514       0.3  0.439863  
6637       0.0  0.440135  
2575       0.0  0.450450  
7336       0.2  0.443746  

[7995 rows x 9 columns]

In [329]:
from sklearn.tree import DecisionTreeRegressor
dtr = DecisionTreeRegressor()
dtr_model = dtr.fit(X_train, y_train)
dtr_cv = cross_val_score(dtr, X_train, y_train,cv = KFold(n_splits=5, random_state=1, shuffle=True), scoring = 'r2')
print("%0.2f accuracy with a standard deviation of %0.2f" % (dtr_cv.mean(), dtr_cv.std()))

0.38 accuracy with a standard deviation of 0.23


In [330]:
from sklearn.metrics import r2_score, mean_absolute_error
import math

dtr_pred = dtr_model.predict(X_test)
scores = r2_score(y_test, dtr_pred )
print('R-Squared value for test score check:', scores )# Check how well the model predict foreign data

# Metrics for Measurement
print("REGRESSOR MODEL PERFORMANCE")
print("Mean absolute error(MAE) =", round(sklearn.metrics.mean_absolute_error(y_test, dtr_pred), 2))
print("Mean squared error(MSE) =", round(sklearn.metrics.mean_squared_error(y_test, dtr_pred), 2))
print("Median absolute error =", round(sklearn.metrics.median_absolute_error(y_test, dtr_pred), 2))
print("Explain variance score =", round(sklearn.metrics.explained_variance_score(y_test, dtr_pred), 2))
print("R2 score =", round(sklearn.metrics.r2_score(y_test, dtr_pred), 2))

R-Squared value for test score check: -300.0442894814536
REGRESSOR MODEL PERFORMANCE
Mean absolute error(MAE) = 7132.06
Mean squared error(MSE) = 100177951.04
Median absolute error = 4260.47
Explain variance score = -148.28
R2 score = -300.04


<u>pls Note this is for academic purpose and this scores are not feasible for a real life application